File: lib/discord.py
from datetime import UTC, datetime

from discord import Embed, Member, Message, PermissionOverwrite, ScheduledEvent
from discord.channel import CategoryChannel, TextChannel
from discord.client import HTTPException
from discord.colour import Color
from discord.enums import EntityType, PrivacyLevel
from discord.interactions import Interaction
from discord.role import Role

from lib.logger import logger
from lib.utils import get_logo


async def send_error(interaction: Interaction, function: str) -> None:
    await interaction.followup.send(
        f"Failed to get the {function}. ❌",
        ephemeral=True,
    )

async def create_events(
        self,
        interaction: Interaction,
        info: dict,
        event_description: str,
        start_time: datetime,
        end_time: datetime,
    ) -> ScheduledEvent | None:
        """
        Create a scheduled event in the discord server.

        Args:
            self: The instance of the cog.
            interaction (Interaction): The application command context.
            info (dict): The information of the CTF.
            event_description (str): The description of the event.
            start_time (datetime): The start time of the event.
            end_time (datetime): The end time of the event.

        Returns:
            Optional[ScheduledEvent]: The created scheduled event or None if failed

        """
        guild = interaction.guild

        if guild is None:
            await send_error(interaction, "event")
            return None

        image_logo = await get_logo(info["logo"])

        try:
            scheduled_event = await guild.create_scheduled_event(
                name=info["title"],
                description=event_description,
                start_time=start_time,
                end_time=end_time,
                entity_type=EntityType.external,
                location=info["url"],
                privacy_level=PrivacyLevel.guild_only,
                image=image_logo,
            )
        except HTTPException as e:
            logger.error(f"Error: {e}")

            if str(e) == "Unsupported image type given":
                return await self.create_events(
                    interaction=interaction,
                    info=info,
                    event_description=event_description,
                    start_time=start_time,
                    end_time=end_time,
                )

            await interaction.followup.send(
                f"Failed to create the event. ❌\n Error: {e}",
                ephemeral=True,
            )
            return None

        return scheduled_event

async def create_channel(
        interaction: Interaction,
        channel_name: str,
        category_id: int,
        role_id: int,
        manager_id: int,
    ) -> TextChannel | None:
        guild = interaction.guild
        if guild is None:
            await send_error(interaction, "channel")
            return None

        category = guild.get_channel(category_id)
        if not isinstance(category, CategoryChannel):
            await send_error(interaction, "category")
            return None

        try:
            channel = await guild.create_text_channel(
                name=channel_name,
                category=category,

            )
            await channel.edit(sync_permissions=False)

            overwrite = PermissionOverwrite()
            overwrite.view_channel = False
            await channel.set_permissions(guild.default_role,overwrite=overwrite)

            manager = guild.get_role(manager_id)
            if manager is None:
                await send_error(interaction, "manager")
                return None
            overwrite = PermissionOverwrite()
            overwrite.view_channel = False
            await channel.set_permissions(manager,overwrite=overwrite)

            role = guild.get_role(role_id)
            if role is None:
                await send_error(interaction, "role")
                return None
            overwrite = PermissionOverwrite()
            overwrite.view_channel = True
            overwrite.send_messages = True
            await channel.set_permissions(role,overwrite=overwrite)
        except HTTPException as e:
            logger.error(f"Error: {e}")
            await interaction.followup.send(
                f"Failed to create the channel or assign the permission. ❌\n Error: {e}",
                ephemeral=True,
            )
            return None
        else:
            return channel

async def create_embed(data: dict, start_time: datetime, end_time: datetime, channel: TextChannel) -> Message:
        description = f"""
**Description:**

{data["description"]}

- **Start Time:** <t:{int(start_time.timestamp())}:f>
- **End Time:** <t:{int(end_time.timestamp())}:f>
- **URL:** {data["url"]}
- **Format:** {data["format"]}
- **Location:** {data["location"]}
- **Weight:** {data["weight"]}
- **Prizes:**\n{data["prizes"]}

"""

        embed = (
            Embed(
                title=data["title"],
                description=description,
                url=data["url"],
                timestamp=datetime.now(tz=UTC),
                color=0xBEBEFE,
            )
            .set_thumbnail(url=data["logo"])
            .set_footer(
                text="Add a reaction to get the ctf role (only if you want to participate). 🙃",
                icon_url=None,
            )
        )

        msg = await channel.send(embed=embed)
        await msg.add_reaction("✅")
        # await msg.pin()
        return msg

async def create_role(interaction: Interaction, name: str) -> Role | None:
        if interaction.guild is None:
            return None

        color = Color.random()
        while color == Color.light_gray():
            color = Color.random()

        return await interaction.guild.create_role(
            name=name,
            color=color,
            mentionable=True,
            hoist=True,
        )


async def check_permission(self,interaction):
    server = await self.bot.database.get_server_by_id(interaction.guild.id)
    role_manager = interaction.guild.get_role(server.role_manager_id)
    if not isinstance(interaction.user, Member) or not role_manager or role_manager not in interaction.user.roles:
        await interaction.followup.send(
            "You don't have the required role to run this command. ❌",
            ephemeral=True,
        )
        return

File: pkg/discord/discord.go
package discordutils

import (
	"ctftime"
	"database"
	"errors"
	"fmt"
	"models"
	"time"

	"github.com/charmbracelet/log"
	"github.com/disgoorg/disgo/bot"
	"github.com/disgoorg/disgo/discord"
	"github.com/disgoorg/disgo/handler"
	"github.com/disgoorg/disgo/rest"
	"github.com/disgoorg/snowflake/v2"
	"golang.org/x/exp/rand"
)

const MaxDescLength = 4096

func CheckPermission(e *handler.CommandEvent) error {
	db := database.GetInstance().Connection()
	sendEphemeral := func(content string) error {
		_, err := e.CreateFollowupMessage(discord.MessageCreate{
			Content: content,
			Flags:   discord.MessageFlagEphemeral,
		})
		return err
	}

	guildID := e.GuildID()
	if guildID == nil {
		return sendEphemeral("Guild not found. ❌")
	}

	guild, err := e.Client().Rest.GetGuild(*guildID, false)
	if err != nil {
		return err
	}

	var server models.Server
	err = server.GetByID(db, *guildID)
	if err != nil {
		return err
	}
	if server == (models.Server{}) {
		return sendEphemeral("Server configuration not found. ❌")
	}

	var roleManager *discord.Role
	for i := range guild.Roles {
		if guild.Roles[i].ID == server.RoleManagerID {
			roleManager = &guild.Roles[i]
			break
		}
	}
	member := e.Member()

	hasRole := false
	if member != nil && roleManager != nil {
		for _, roleID := range member.RoleIDs {
			if roleID == roleManager.ID {
				hasRole = true
				break
			}
		}
	}

	if member == nil || roleManager == nil || !hasRole {
		return sendEphemeral("You don't have the required role to run this command. ❌")
	}

	return nil
}

func CreateEvents(client *bot.Client, guildID *snowflake.ID, info ctftime.Event, eventDescription string, startTime time.Time, endTime time.Time) (discord.GuildScheduledEvent, error) {
	var image *discord.Icon
	if info.Logo != "" {
		imageLogo, imageType, err := ctftime.GetLogo(info.Logo)
		if err != nil {
			return discord.GuildScheduledEvent{}, err
		}
		log.Debug("Creating event with logo", "imageType", imageType)

		image = &discord.Icon{
			Data: imageLogo,
			Type: discord.IconType(imageType),
		}
	}

	scheduledEvent, err := client.Rest.CreateGuildScheduledEvent(*guildID, discord.GuildScheduledEventCreate{
		Name:               info.Title,
		Description:        eventDescription,
		ScheduledStartTime: startTime,
		ScheduledEndTime:   &endTime,
		EntityType:         discord.ScheduledEventEntityTypeExternal,
		PrivacyLevel:       discord.ScheduledEventPrivacyLevelGuildOnly,
		Image:              image,
		EntityMetaData: &discord.EntityMetaData{
			Location: info.URL,
		},
	})
	if err != nil {
		if restErr, ok := err.(*rest.Error); ok && restErr.Message == "Unsupported image type given" {
			return CreateEvents(client, guildID, info, eventDescription, startTime, endTime)
		}
		return discord.GuildScheduledEvent{}, err
	}

	return *scheduledEvent, nil
}

func CreateChannel(client *bot.Client, guildID *snowflake.ID, channelName string, categoryID snowflake.ID, roleID snowflake.ID, managerID snowflake.ID) (*discord.GuildChannel, error) {
	log.Debug("CreateChannel called with",
		"channelName", channelName,
		"categoryID", categoryID,
		"roleID", roleID,
		"managerID", managerID,
	)

	log.Debug("Fetching category channel", "categoryID", categoryID)
	category, err := client.Rest.GetChannel(categoryID)
	if err != nil {
		log.Debug("failed to fetch category channel", "err", err)
	}
	if err != nil || category.Type() != discord.ChannelTypeGuildCategory {
		log.Debug("Category channel is invalid or not a category", "type", category.Type())
		return nil, errors.New("invalid category")
	}

	var everyoneID *snowflake.ID
	roles, err := client.Rest.GetRoles(*guildID)
	if err != nil {
		log.Debug("failed to fetch roles", "err", err)
		return nil, err
	}
	for _, role := range roles {
		if role.Name == "@everyone" {
			log.Debug("Found @everyone role", "roleID", role.ID)
			everyoneID = &role.ID
			break
		}
	}

	if everyoneID == nil {
		log.Debug("Could not find @everyone role in guild", "guildID", *guildID)
		return nil, errors.New("could not find @everyone role in guild")
	}

	log.Debug("Creating guild text channel",
		"channelName", channelName,
		"categoryID", categoryID,
	)
	channel, err := client.Rest.CreateGuildChannel(*guildID, discord.GuildTextChannelCreate{
		Name:     channelName,
		ParentID: categoryID,
		PermissionOverwrites: []discord.PermissionOverwrite{
			discord.RolePermissionOverwrite{
				RoleID: *everyoneID,
				Deny:   discord.PermissionViewChannel,
			},
			discord.RolePermissionOverwrite{
				RoleID: roleID,
				Allow:  discord.PermissionViewChannel | discord.PermissionSendMessages,
			},
			discord.RolePermissionOverwrite{
				RoleID: managerID,
				Deny:   discord.PermissionViewChannel,
			},
		},
	})
	if err != nil {
		log.Debug("failed to create guild channel", "err", err)
		return nil, err
	}

	log.Debug("Successfully created channel",
		"channelName", channel.Name(),
		"channelID", channel.ID(),
	)
	return &channel, nil
}

func CreateEmbed(client *bot.Client, data ctftime.Event, startTime time.Time, endTime time.Time, channel discord.Channel) (*discord.Message, error) {
	description := fmt.Sprintf(`
**Description:**

%s

- **Start Time:** <t:%d:f>
- **End Time:** <t:%d:f>
- **URL:** %s
- **Format:** %s
- **Location:** %s
- **Weight:** %f
- **Prizes:**
%s

`, data.Description, int(startTime.Unix()), int(endTime.Unix()), data.URL, data.Format, data.Location, data.Weight, data.Prizes)

	t := time.Now()
	embed := discord.Embed{
		Title:       data.Title,
		Description: description,
		URL:         data.URL,
		Timestamp:   &t,
		Color:       0xBEBEFE,
		Thumbnail: &discord.EmbedResource{
			URL: data.Logo,
		},
		Footer: &discord.EmbedFooter{
			Text: "Add a reaction to get the ctf role (only if you want to participate). 🙃",
		},
	}

	message, err := client.Rest.CreateMessage(channel.ID(), discord.MessageCreate{
		Embeds: []discord.Embed{embed},
	})
	if err != nil {
		return nil, err
	}

	// Add reaction
	err = client.Rest.AddReaction(channel.ID(), message.ID, "✅")
	if err != nil {
		return nil, err
	}

	return message, nil
}

func CreateRole(client *bot.Client, guildID *snowflake.ID, name string) (*discord.Role, error) {
	if guildID == nil {
		return nil, nil
	}

	color := rand.Intn(0xFFFFFF)
	for color == 0xBEBEFE {
		color = rand.Intn(0xFFFFFF)
	}

	role, err := client.Rest.CreateRole(*guildID, discord.RoleCreate{
		Name:        name,
		Color:       color,
		Mentionable: true,
		Hoist:       true,
	})
	return role, err
}
